include/mbedtls/inst.h

```diff
@@ -0,0 +1,108 @@
+/**
+ * \file inst.h
+ *
+ * \brief Instrumentation macros.
```

Comment:
No gratuitous abbreviations please. If this is about _instrumentation_, call it _instrumentation_, not _inst_ (instance? installation? instruction? instant?). But given what the macros do, I think _trace_ would be a better name.

Reply:
@gilles-peskine-arm Gilles,
What wrapper are you proposing? The native -finstrument-functions wrapper in GCC for example does not provide a method for reporting arguments. I would much rather NOT have all of this code in the library. Can you provide an example of how these functions may be wrapped at link time?

Reply:
Circling back on this comment, "Wouldn't it be enough to wrap those functions by print-and-call functions at link time?" Can you point me to an example of what you are proposing? I would also like to avoid extra macros littering the code if there is a better way to trace functions and function arguments in a reasonably performant manner (e.g., not stepping through GDB and dumping the stack after every next line).

Reply:
I haven't checked very closely: are there places where `mbedtls_mod_f()` calls `mbedtls_mod_g()`, and you want to have a trace both for f and g? If so then I don't have an alternative proposal. If not, then one way to generate traces is to
```
#define mod_f mod_f_real
```
when building `mod.c`, and link with
```
mod_f(args) {
    trace(args);
    return mod_f_real(args);
}
```
The wrappers are preferably automatically generated. FFI tools might help with that but I don't have a concrete tool to propose.

Reply:
Perfect timing: I was just reading the DEF CON 29 papers and found a dynamic parasitic tracing tool called [FRIDA](https://frida.re/docs/installation/) which may actually be a solution to this tracing problem. I am going to investigate this and report back if it can obviate the need for tracing.